### ISSUE TITLE
TOTP Seed (KeePassXC/KeeTrayTOTP) Compatibility?

### ISSUE THREAD

**Issue by BoffinBrain** [tiuub/KeeOtp2#48]
I have given KeeOtp2 a try and I'm very impressed by how intuitive and feature-rich this plugin is! However, I have noticed that it uses its own custom field type which isn't the same as those used by the classic TrayTOTP.

TrayTOTP and its successor KeePass plugin, KeeTrayTOTP, uses the fields `TOTP Seed` and `TOTP Settings`. Due to its popularity, KeePassXC and KeePass2Android (and possibly others) also support these fields by default.

I know KeeOtp has its own backwards compatibility to consider as well, but do you have any plans to support this other widely-used field format?

**Comment by maddes-b** [comment]
[KeeTrayTOTP](https://github.com/KeeTrayTOTP/KeeTrayTOTP) is "outdated".
KeePass got its own TOTP implementation, just convert the attribute fields.
KeePassXC has a little different TOTP implementation (using "otpauth" string) but also supports KeePass implementation sind 2.7.10 (2025-03).
The rest of KeeTrayTOTP is already inside KeeOtp2.

**Comment by maddes-b** [comment]
KeeTrayTOTP is "outdated" as KeePass got its [own TOTP implementation](https://keepass.info/help/base/placeholders.html#hmacotp), and KeeTrayTOTP was not updated to this.
Still KeeTrayTOTP does its job well.

My path was to switch to KeePass2 definitions, so no need for plugins at work, and use plugin KeeOtp2 for Steam (plus to show TOTP QR code).
The KeePass2 definitions are also supported in KeePassXC since 2.7.10.
The switch to KeePass2 is quite simple, while the switch for Steam is more complicated.

Preparation:
* Close KeePass, make a backup copy of the KDBX files before manipulating them and set the backup copies to read-only.
* Restart KeePass.

Convert regular TOTPs from KeeTrayTOTP to KeePass2:
* First check the attribute field `TOTP Settings` which holds the two values "period" and "length" separated by semicolon.
  If the second value "length" is not a number, then it is not a regular TOTP and has to be handled differently, e.g. `S` is for Steam.
  Otherwise continue as following.
* The attribute field `TOTP Seed` must be copied/renamed to `TimeOtp-Secret-Base32`. Base32 is the only encoding supported by KeeTrayTOTP.
* The attribute field `TOTP Settings` must be split/copied to `TimeOtp-Period` and `TimeOtp-Length`.
  * First value "period" must only be copied if it differs from 30 (TOTP default).
  * Second value "length" must only be copied if it differs from 6 (TOTP default).
* (optional) Remove attribute field `TOTP Settings`.
* Last but not least remove spaces inside `TimeOtp-Secret-Base32` field attributes, as they can lead to failures.

Convert Steam TOTPs from KeeTrayTOTP to KeeOtp2:
* Have a text editor open.
* First check the attribute field `TOTP Settings` which holds the two values "period" and "length" separated by semicolon.
  If the second value "length" is not `S` for Steam, then it is not a Steam TOTP and has to be handled differently.
  Otherwise continue as following.
* Add field attribute `convert_steam` with definition `key={S:TOTP Seed}&type=Steam&size=5` (for KeeOtp2, not compatible with KeePassXC, see https://github.com/tiuub/KeeOtp2/issues/55) . Save and close entry.
  * **Update 2025-03-21**: KeeOtp2 1.6.0 supports Steam as an otppath URI like done by KeePassXC. `otpauth://totp/{USERNAME}?secret={S:TOTP Seed}&digits=5&encoder=steam`
* Choose menu "Entry > Other Data > Copy convert_steam" and temporary paste string into text editor.
* Add field attribute `otp` and copy the string from the editor as its value.
* (optional) Remove attribute fields `TOTP Seed`, `TOTP Settings` and `convert_steam`.

For converting the regular TOTPs from KeeTrayTOTP to KeePass2 I used the [plugin KPFieldsAdminConsole](https://keepass.info/plugins.html#kpfieldsadmin).
This should also be achievable with [XML Replace](https://keepass.info/help/v2/xml_replace.html), but I prefer the GUI from the plugin.

Preparation:
* Close KeePass and install the [plugin KPFieldsAdminConsole](https://keepass.info/plugins.html#kpfieldsadmin).
* Restart KeePass.

 Convert Step 1
* Choose menu "Tools > Fields Admin Console".
* First check mark the field `TOTP Seed` on the left.
* Click button "Rename" for fields on the left and convert field name from `TOTP Seed` to `TimeOtp-Secret-Base32`.
  That will also modify Steam TOTPs entries, so for the field attribute `convert_steam` the definition `key={S:TimeOtp-Secret-Base32}&type=Steam&size=5` has to be used.
  * **Update 2025-03-21**: KeeOtp2 1.6.0 supports Steam as an otppath URI like done by KeePassXC. `otpauth://totp/{USERNAME}?secret={S:TimeOtp-Secret-Base32}&digits=5&encoder=steam`
* Close Fields Admin Console.

 Convert Step 2
* Choose menu "Tools > Fields Admin Console".
* Click on field name `TOTP Settings` on the left, but do not check mark it.
* Filter values on the right for `30;6` only (TOTP defaults), so only that value is check marked.
* Click button "Delete" for values on the right - done for all Standard TOTPs.
* The remaining entries shown in the middle should be noted down and have to be changed manually like described before.
* Close Fields Admin Console.

Check for spaces
* Choose menu "Tools > Fields Admin Console".
* Click on field name `TOTP Settings` on the left, but do not check mark it.
* Filter values on the right for spaces, so only those values are check marked.
* The entries shown in the middle should be noted down and have to be changed manually to remove the spaces.

Hope this helps
Maddes

**Comment by BoffinBrain** [comment]
Thanks for the detailed steps, Maddes!

It took me a few reads to understand it, but I was able to do a migration with the KPFieldsAdminConsole plugin, fix up the seeds that contained spaces, and verify that KeeOtp2 could handle Steam OTPs.

The only thing that KeeOtp2 doesn't seem to do, which KeeTrayTOTP could, is show TOTP codes in the entries list as a custom column, to quickly see which entries had 2FA enabled. I guess I can use KPFieldsAdminConsole for the rare occasions where I need to do that.

I guess this issue ticket is obsolete so I'll close it. Cheers!

**Comment by maddes-b** [comment]
You are right, that custom column is not available in KeePass2 and KeeOtp2.
I never came across that column. Wouldn't enable it, as I use my laptop also in crowded spaces (work, cafe, etc.)

**Comment by maddes-b** [comment]
@BoffinBrain Just released KeeOtp2 1.6.0 supports Steam as an otppath URI like done by KeePassXC. Still has [a problem](https://github.com/tiuub/KeeOtp2/issues/59) with URI encoded secrets.

**Comment by BoffinBrain** [comment]
1.6.0 seems to be working fine for regular and Steam OTPs. I'm not really sure what an otppath URI is, but I have mine stored in the `otp` field like your guide stated.

**Comment by maddes-b** [comment]
If the value of `otp` begins with `otpauth://` then it is an URI.
If the secret has one or more `%3D` at the end, then you run into this problem with KeeOtp2.
This can be the case, when TOTP is created with KeePassXC.
See https://github.com/tiuub/KeeOtp2/pull/61
